src/cosalib/qemuvariants.py

```diff
@@ -0,0 +1,220 @@
+"""
+Provides a base abstration for building images.
+"""
+
+import logging as log
+import os.path
+import shutil
+import sys
+import tarfile
+
+cosa_dir = os.path.dirname(os.path.abspath(__file__))
+sys.path.insert(0, f"{cosa_dir}/cosalib")
+sys.path.insert(0, cosa_dir)
+
+from cosalib.build import _Build
+from cosalib.cmdlib import (
+    get_basearch,
+    run_verbose,
+    sha256sum_file
+)
+
+
+# BASEARCH is the current machine architecture
+BASEARCH = get_basearch()
+
+# Variant are disk types that are derived from qemu images.
+# To define new variants that use the QCOW2 disk image, simply,
+# add its definition below:
+VARIANTS = {
+    "aws": {
+        "image_format": "vmdk",
+        "image_suffix": "vmdk",
+        "platform": "aws",
+        "convert_options": {
+            '-o': 'adapter_type=lsilogic,subformat=streamOptimized,compat6'
+        },
+    },
+    "aliyun": {
+        "image_format": "qcow2",
+        "platform": "aliyun",
+    },
+    "azure": {
+        "image_format": "vpc",
+        "image_suffix": "vhd",
+        "platform": "azure",
+    },
+    "openstack": {
+        "image_format": "qcow2",
+        "platform": "openstack",
+    },
+    "gcp": {
+        "image_format": "raw",
+        "platform": "gcp",
+        "image_suffix": "tar.gz",
+        "tar_member": True,
+    },
+    "vmware_vmdk": {
+        "image_format": "vmdk",
+        "image_suffix": "vmdk",
+        "platform": "vmware",
+        "convert_options":  {
+            '-o': 'adapter_type=lsilogic,subformat=streamOptimized,compat6'
+        }
+    }
+}
+
+
+class ImageError(Exception):
```

Comment:
:+1:

Reply:
Once this and the other PR lands, I have this:
https://github.com/darkmuggle/coreos-assembler/commit/1833b4d4e9361177464252300fe7482a2f00122f

That will improve by have a "BuildExistsError" making the builds idempotent.